include/boost/beast/http/impl/read.ipp

```diff
@@ -414,32 +413,27 @@ operator()(
     std::size_t bytes_transferred)
 {
     auto& d = *d_;
-    switch(d.state)
+    BOOST_ASIO_CORO_REENTER(d)
     {
-    case 0:
-        d.state = 1;
-
-    do_read:
-        return async_read_some(
-            d.s, d.b, d.p, std::move(*this));
-
-    case 1:
-    case 2:
-        if(ec)
-            goto upcall;
-        d.bytes_transferred +=
-            bytes_transferred;
-        if(d.p.is_done())
+        while (!d.p.is_done())
         {
-            d.m = d.p.release();
-            goto upcall;
+            BOOST_ASIO_CORO_YIELD async_read_some(
+                d.s, d.b, d.p, std::move(*this));
+            if (ec)
+            {
+                break;
+            }
+            d.bytes_transferred += bytes_transferred;
+
         }
-        d.state = 2;
-        goto do_read;
     }
-upcall:
-    bytes_transferred = d.bytes_transferred;
-    d_.invoke(ec, bytes_transferred);
+    if (d.is_complete())
```

Comment:
This performs an unnecessary check on every continuation. Consider this instead:
```
    }
    return;
upcall:
    d_.invoke(ec, bytes_transferred);
}
```

And use `goto upcall` to invoke the final completion handler